File: dotnet/src/Ode.Adapter/Cow.cs
using System.Text.Json.Nodes;

namespace Ode.Adapter;

/// <summary>COW subset helpers — the scoped Clinical Order Workflows layer.
/// Faithful port of cow.py. Builds the COW/FHIR artifacts the bridge applies on every
/// reply: the dental businessStatus, Task.output wrapping, the provisional dental
/// profiles, the ODEMedicationList, and the non-document reply resources (Appointment
/// from PCC-60, Communication from PCC-61). Resources embed a `_fullUrl` (stripped when
/// wrapped into a Bundle entry), mirroring the Python dict shape.</summary>
public static class Cow
{
    private static string Urn() => $"urn:uuid:{Guid.NewGuid()}";

    private const string UsCorePractitioner =
        "http://hl7.org/fhir/us/core/StructureDefinition/us-core-practitioner";
    private const string UsCoreOrganization =
        "http://hl7.org/fhir/us/core/StructureDefinition/us-core-organization";

    /// <summary>CodeableConcept for Task.businessStatus from the dental value set.</summary>
    public static JsonObject BusinessStatusConcept(string code)
    {
        var display = Config.CowBusinessStatus.TryGetValue(code, out var d) ? d : code;
        return new JsonObject
        {
            ["coding"] = new JsonArray
            { new JsonObject { ["system"] = Config.CowBusinessStatusSystem, ["code"] = code, ["display"] = display } },
            ["text"] = display,
        };
    }

    private static JsonObject? Concept(string system, IReadOnlyDictionary<string, string> table, string? code)
    {
        if (string.IsNullOrEmpty(code)) return null;
        if (table.TryGetValue(code, out var display))
            return new JsonObject
            {
                ["coding"] = new JsonArray
                { new JsonObject { ["system"] = system, ["code"] = code, ["display"] = display } },
                ["text"] = display,
            };
        return new JsonObject { ["text"] = code };   // unrecognized code -> carry the text
    }

    public static JsonObject? DeclineReasonConcept(string? code) =>
        Concept(Config.DeclineReasonSystem, Config.DeclineReasons, code);

    public static JsonObject? AppointmentTypeConcept(string? code) =>
        Concept(Config.AppointmentTypeSystem, Config.AppointmentTypes, code);

    public static JsonObject? NoshowReasonConcept(string? code) =>
        Concept(Config.NoshowReasonSystem, Config.NoshowReasons, code);

    public static JsonObject? ClearanceConcept(string? code) =>
        Concept(Config.ClearanceSystem, Config.ClearanceDispositions, code);

    /// <summary>Build the accepting provider's identity for Task.owner (PCC-56 accept).
    /// Returns (resources, ownerRef): resources to write + the PractitionerRole fullUrl.</summary>
    public static (List<JsonObject> resources, string? ownerRef) BuildOwnerRole(JsonObject? provider)
    {
        provider ??= new JsonObject();
        var any = new[] { "name", "npi", "specialty", "organization" }
            .Any(k => !string.IsNullOrEmpty(JsonX.Str(provider[k])));
        if (!any) return (new List<JsonObject>(), null);

        var pract = new JsonObject
        {
            ["resourceType"] = "Practitioner",
            ["_fullUrl"] = Urn(),
            ["meta"] = new JsonObject { ["profile"] = new JsonArray { UsCorePractitioner } },
            ["name"] = new JsonArray { new JsonObject { ["text"] = JsonX.Str(provider["name"]) ?? "Accepting Provider" } },
        };
        var npi = JsonX.Str(provider["npi"]);
        if (!string.IsNullOrEmpty(npi))
            pract["identifier"] = new JsonArray { new JsonObject { ["system"] = Config.SysNpi, ["value"] = npi } };
        var resources = new List<JsonObject> { pract };

        JsonObject? org = null;
        var orgName = JsonX.Str(provider["organization"]);
        if (!string.IsNullOrEmpty(orgName))
        {
            org = new JsonObject
            {
                ["resourceType"] = "Organization",
                ["_fullUrl"] = Urn(),
                ["name"] = orgName,
                ["meta"] = new JsonObject { ["profile"] = new JsonArray { UsCoreOrganization } },
            };
            resources.Add(org);
        }
        var role = new JsonObject
        {
            ["resourceType"] = "PractitionerRole",
            ["_fullUrl"] = Urn(),
            ["practitioner"] = new JsonObject { ["reference"] = JsonX.Str(pract["_fullUrl"]) },
        };
        if (org != null) role["organization"] = new JsonObject { ["reference"] = JsonX.Str(org["_fullUrl"]) };
        var spec = JsonX.Str(provider["specialty"]);
        if (!string.IsNullOrEmpty(spec))
            role["specialty"] = new JsonArray { new JsonObject
            {
                ["coding"] = new JsonArray { new JsonObject { ["system"] = Config.SysNucc, ["display"] = spec } },
                ["text"] = spec,
            } };
        resources.Add(role);
        return (resources, JsonX.Str(role["_fullUrl"]));
    }

    /// <summary>A single interim finding Observation for PCC-59 (in addition to the note).</summary>
    public static JsonObject InterimObservation(string referralId, string? patientRef,
                                                string? finding, string? value = null)
    {
        var obs = new JsonObject
        {
            ["resourceType"] = "Observation",
            ["_fullUrl"] = Urn(),
            ["identifier"] = new JsonArray { new JsonObject { ["system"] = Config.SysReferralId, ["value"] = referralId } },
            ["status"] = "preliminary",
            ["code"] = new JsonObject { ["text"] = string.IsNullOrEmpty(finding) ? "Interim finding" : finding },
        };
        if (patientRef != null) obs["subject"] = new JsonObject { ["reference"] = patientRef };
        if (!string.IsNullOrEmpty(value)) obs["valueString"] = value;
        return obs;
    }

    /// <summary>A coded clearance/disposition Observation for PCC-57 outcome.</summary>
    public static JsonObject ClearanceObservation(string referralId, string? patientRef, string disposition)
    {
        var concept = ClearanceConcept(disposition) ?? new JsonObject { ["text"] = disposition };
        var obs = new JsonObject
        {
            ["resourceType"] = "Observation",
            ["_fullUrl"] = Urn(),
            ["identifier"] = new JsonArray { new JsonObject { ["system"] = Config.SysReferralId, ["value"] = referralId } },
            ["status"] = "final",
            ["code"] = new JsonObject
            {
                ["coding"] = new JsonArray { new JsonObject
                { ["system"] = Config.SysLoinc, ["code"] = Config.ClearanceLoinc, ["display"] = "Referral clearance / disposition" } },
                ["text"] = "Referral clearance / disposition",
            },
            ["valueCodeableConcept"] = concept,
        };
        if (patientRef != null) obs["subject"] = new JsonObject { ["reference"] = patientRef };
        return obs;
    }

    /// <summary>The ODEMedicationList (`List`) aggregating the patient's current medications.
    /// Base-FHIR List (status=current, mode=snapshot, code=LOINC 10160-0) referencing the
    /// US Core MedicationRequests; referenced from ServiceRequest.supportingInfo.</summary>
    public static JsonObject BuildMedicationList(string referralId, string? patientRef, IEnumerable<string> medUrls)
    {
        var entry = new JsonArray();
        foreach (var u in medUrls) entry.Add(new JsonObject { ["item"] = new JsonObject { ["reference"] = u } });
        var lst = new JsonObject
        {
            ["resourceType"] = "List",
            ["_fullUrl"] = Urn(),
            ["meta"] = new JsonObject { ["profile"] = new JsonArray { Config.ProfileOdeMedicationList } },
            ["identifier"] = new JsonArray { new JsonObject { ["system"] = Config.SysReferralId, ["value"] = referralId } },
            ["status"] = "current",
            ["mode"] = "snapshot",
            ["title"] = "Medication List",
            ["code"] = new JsonObject
            {
                ["coding"] = new JsonArray { new JsonObject
                { ["system"] = Config.SysLoinc, ["code"] = Config.MedListLoinc, ["display"] = "History of Medication use Narrative" } },
                ["text"] = "Medication List",
            },
            ["entry"] = entry,
        };
        if (patientRef != null) lst["subject"] = new JsonObject { ["reference"] = patientRef };
        return lst;
    }

    /// <summary>Wrap result resources as Task.output entries (reference by fullUrl/id).</summary>
    public static JsonArray TaskOutput(IEnumerable<JsonObject>? resources)
    {
        var outArr = new JsonArray();
        foreach (var res in resources ?? Enumerable.Empty<JsonObject>())
        {
            var reference = JsonX.Str(res["_fullUrl"]);
            if (string.IsNullOrEmpty(reference))
            {
                var id = JsonX.Str(res["id"]);
                reference = id != null ? $"{JsonX.Str(res["resourceType"])}/{id}" : null;
            }
            if (string.IsNullOrEmpty(reference)) continue;
            outArr.Add(new JsonObject
            {
                ["type"] = new JsonObject { ["text"] = JsonX.Str(res["resourceType"]) },
                ["valueReference"] = new JsonObject { ["reference"] = reference },
            });
        }
        return outArr;
    }

    /// <summary>Stamp provisional ODE dental profiles on dental-flavored resources in place.</summary>
    public static void ApplyDentalProfiles(IEnumerable<JsonObject>? resources)
    {
        foreach (var res in resources ?? Enumerable.Empty<JsonObject>())
        {
            var rtype = JsonX.Str(res["resourceType"]);
            if (rtype == "Procedure" && IsCdt(res))
                AddProfile(res, Config.ProfileDentalProcedure);
            else if (rtype == "Observation" && (JsonX.Bool(res["_dental_perio"]) || ProfileHas(res, "perio")))
                AddProfile(res, Config.ProfilePerioObservation);
        }
    }

    /// <summary>COW reply resource for PCC-60 Appointment Notification.</summary>
    public static JsonObject BuildAppointment(string referralId, string? patientRef, string? taskRef,
        string? start, string status = "booked", string? end = null, string? location = null,
        string? provider = null, string? apptType = null)
    {
        var appt = new JsonObject
        {
            ["resourceType"] = "Appointment",
            ["_fullUrl"] = Urn(),
            ["identifier"] = new JsonArray { new JsonObject { ["system"] = Config.SysReferralId, ["value"] = referralId } },
            ["status"] = status,
        };
        if (!string.IsNullOrEmpty(start)) appt["start"] = Iso(start);
        if (!string.IsNullOrEmpty(end)) appt["end"] = Iso(end);
        var typeCc = AppointmentTypeConcept(apptType);
        if (typeCc != null) appt["appointmentType"] = typeCc;
        if (!string.IsNullOrEmpty(location)) appt["description"] = location;
        var parts = new JsonArray();
        if (patientRef != null)
            parts.Add(new JsonObject { ["actor"] = new JsonObject { ["reference"] = patientRef }, ["status"] = "accepted" });
        if (!string.IsNullOrEmpty(provider))
            parts.Add(new JsonObject { ["actor"] = new JsonObject { ["display"] = provider }, ["status"] = "accepted" });
        if (parts.Count > 0) appt["participant"] = parts;
        if (taskRef != null) appt["basedOn"] = new JsonArray { new JsonObject { ["reference"] = taskRef } };
        return appt;
    }

    /// <summary>COW reply resource for PCC-61 No-Show Notification.</summary>
    public static JsonObject BuildCommunication(string referralId, string? patientRef, string? taskRef,
        string reason, string? reasonCode = null, string? reschedule = null)
    {
        var payloads = new JsonArray { new JsonObject { ["contentString"] = reason } };
        if (!string.IsNullOrEmpty(reschedule))
            payloads.Add(new JsonObject { ["contentString"] = $"Reschedule: {reschedule}" });
        var comm = new JsonObject
        {
            ["resourceType"] = "Communication",
            ["_fullUrl"] = Urn(),
            ["identifier"] = new JsonArray { new JsonObject { ["system"] = Config.SysReferralId, ["value"] = referralId } },
            ["status"] = "completed",
            ["reasonCode"] = new JsonArray { NoshowReasonConcept(reasonCode) ?? new JsonObject { ["text"] = reason } },
            ["payload"] = payloads,
        };
        if (patientRef != null) comm["subject"] = new JsonObject { ["reference"] = patientRef };
        if (taskRef != null) comm["partOf"] = new JsonArray { new JsonObject { ["reference"] = taskRef } };
        return comm;
    }

    /// <summary>A COW Task resource snapshot for the harness inbox / export.</summary>
    public static JsonObject TaskSnapshot(string referralId, string? taskId, string status,
        string businessStatus, string? focusRef = null, string? patientRef = null,
        JsonArray? outputs = null, JsonNode? owner = null, JsonObject? statusReason = null,
        string? note = null, string? periodEnd = null)
    {
        var task = new JsonObject
        {
            ["resourceType"] = "Task",
            ["meta"] = new JsonObject { ["profile"] = new JsonArray { Config.ProfileCowTask } },
            ["identifier"] = new JsonArray { new JsonObject { ["system"] = Config.SysReferralId, ["value"] = referralId } },
            ["status"] = status,
            ["intent"] = "order",
            ["businessStatus"] = BusinessStatusConcept(businessStatus),
            ["code"] = new JsonObject
            {
                ["coding"] = new JsonArray { new JsonObject
                { ["system"] = "http://hl7.org/fhir/CodeSystem/task-code", ["code"] = "fulfill" } }
            },
        };
        if (!string.IsNullOrEmpty(taskId)) task["id"] = taskId;
        if (focusRef != null) task["focus"] = new JsonObject { ["reference"] = focusRef };
        if (patientRef != null) task["for"] = new JsonObject { ["reference"] = patientRef };
        if (owner != null)
        {
            var s = JsonX.Str(owner);
            task["owner"] = s != null ? new JsonObject { ["reference"] = s } : owner.DeepClone();
        }
        if (statusReason != null) task["statusReason"] = JsonX.Clone(statusReason);
        if (!string.IsNullOrEmpty(note)) task["note"] = new JsonArray { new JsonObject { ["text"] = note } };
        if (!string.IsNullOrEmpty(periodEnd))
            task["restriction"] = new JsonObject { ["period"] = new JsonObject { ["end"] = periodEnd } };
        if (outputs != null) task["output"] = (JsonArray)outputs.DeepClone();
        return task;
    }

    // --------------------------------------------------------------------- //
    /// <summary>Accept an HL7 v2 timestamp (YYYYMMDDHHMMSS) or pass through ISO-8601.</summary>
    public static string Iso(string? value)
    {
        var v = (value ?? "").Trim();
        if (v.Length >= 8 && v.All(char.IsDigit))
        {
            var outp = $"{v[..4]}-{v.Substring(4, 2)}-{v.Substring(6, 2)}";
            if (v.Length >= 12)
            {
                var sec = v.Length >= 14 ? v.Substring(12, 2) : "00";
                outp += $"T{v.Substring(8, 2)}:{v.Substring(10, 2)}:{sec}Z";
            }
            return outp;
        }
        return v;
    }

    private static bool IsCdt(JsonObject res)
    {
        var coding = JsonX.Arr(JsonX.Obj(res["code"])?["coding"]);
        if (coding == null) return false;
        foreach (var c in coding)
            if (JsonX.Str((c as JsonObject)?["system"]) == Config.SysCdt) return true;
        return false;
    }

    private static bool ProfileHas(JsonObject res, string needle)
    {
        var profs = JsonX.Arr(JsonX.Obj(res["meta"])?["profile"]);
        if (profs == null) return false;
        foreach (var p in profs)
        {
            var s = JsonX.Str(p);
            if (s != null && s.ToLowerInvariant().Contains(needle)) return true;
        }
        return false;
    }

    private static void AddProfile(JsonObject res, string profile)
    {
        var meta = JsonX.Obj(res["meta"]);
        if (meta == null) { meta = new JsonObject(); res["meta"] = meta; }
        var profs = JsonX.Arr(meta["profile"]);
        if (profs == null) { profs = new JsonArray(); meta["profile"] = profs; }
        if (!profs.Any(p => JsonX.Str(p) == profile)) profs.Add(profile);
    }
}

File: python/ode_adapter/cow.py
"""COW subset helpers — the scoped Clinical Order Workflows layer.

Everything here is *scoped to what 360X supports* (see
spec/mapping/360x-cow-crosswalk.md). It builds the COW/FHIR artifacts the bridge
applies on every reply: the dental `businessStatus`, `Task.output` wrapping, the
provisional dental profiles, and the non-document reply resources (Appointment from
PCC-60, Communication from PCC-61). Keeping it in one module makes the COW subset
auditable against the crosswalk.
"""
from __future__ import annotations

import uuid

from . import config


def _urn() -> str:
    return f"urn:uuid:{uuid.uuid4()}"


def business_status_concept(code: str) -> dict:
    """CodeableConcept for Task.businessStatus from the dental value set."""
    return {"coding": [{"system": config.COW_BUSINESS_STATUS_SYSTEM, "code": code,
                        "display": config.COW_BUSINESS_STATUS.get(code, code)}],
            "text": config.COW_BUSINESS_STATUS.get(code, code)}


def _concept(system: str, table: dict, code: str | None) -> dict | None:
    """CodeableConcept from a provisional value set; falls back to free text."""
    if not code:
        return None
    display = table.get(code)
    if display:
        return {"coding": [{"system": system, "code": code, "display": display}],
                "text": display}
    return {"text": code}   # unrecognized code -> carry the text (degraded)


def decline_reason_concept(code: str | None) -> dict | None:
    """Task.statusReason for a PCC-56 decline (coded reason)."""
    return _concept(config.DECLINE_REASON_SYSTEM, config.DECLINE_REASONS, code)


def appointment_type_concept(code: str | None) -> dict | None:
    return _concept(config.APPOINTMENT_TYPE_SYSTEM, config.APPOINTMENT_TYPES, code)


def noshow_reason_concept(code: str | None) -> dict | None:
    return _concept(config.NOSHOW_REASON_SYSTEM, config.NOSHOW_REASONS, code)


def clearance_concept(code: str | None) -> dict | None:
    return _concept(config.CLEARANCE_SYSTEM, config.CLEARANCE_DISPOSITIONS, code)


def build_owner_role(provider: dict | None) -> tuple[list[dict], str | None]:
    """Build the accepting provider's identity for Task.owner (PCC-56 accept).

    Returns (resources, owner_ref): the PractitionerRole (+ Practitioner + optional
    Organization) to write, and the PractitionerRole fullUrl to set as Task.owner.
    An empty provider yields ([], None) so accept still works without an identity.
    """
    provider = provider or {}
    if not any(provider.get(k) for k in ("name", "npi", "specialty", "organization")):
        return [], None
    pract = {"resourceType": "Practitioner", "_fullUrl": _urn(),
             "meta": {"profile": [
                 "http://hl7.org/fhir/us/core/StructureDefinition/us-core-practitioner"]},
             "name": [{"text": provider.get("name") or "Accepting Provider"}]}
    if provider.get("npi"):
        pract["identifier"] = [{"system": config.SYS_NPI, "value": provider["npi"]}]
    resources = [pract]
    org = None
    if provider.get("organization"):
        org = {"resourceType": "Organization", "_fullUrl": _urn(),
               "name": provider["organization"],
               "meta": {"profile": [
                   "http://hl7.org/fhir/us/core/StructureDefinition/us-core-organization"]}}
        resources.append(org)
    role = {"resourceType": "PractitionerRole", "_fullUrl": _urn(),
            "practitioner": {"reference": pract["_fullUrl"]}}
    if org:
        role["organization"] = {"reference": org["_fullUrl"]}
    if provider.get("specialty"):
        role["specialty"] = [{"coding": [{"system": config.SYS_NUCC,
                                          "display": provider["specialty"]}],
                              "text": provider["specialty"]}]
    resources.append(role)
    return resources, role["_fullUrl"]


def interim_observation(referral_id: str, patient_ref: str | None,
                        finding: str, value: str | None = None) -> dict:
    """A single interim finding Observation for PCC-59 (in addition to the note)."""
    obs = {"resourceType": "Observation", "_fullUrl": _urn(),
           "identifier": [{"system": config.SYS_REFERRAL_ID, "value": referral_id}],
           "status": "preliminary",
           "code": {"text": finding or "Interim finding"}}
    if patient_ref:
        obs["subject"] = {"reference": patient_ref}
    if value:
        obs["valueString"] = value
    return obs


def clearance_observation(referral_id: str, patient_ref: str | None,
                          disposition: str) -> dict:
    """A coded clearance/disposition Observation for PCC-57 outcome."""
    concept = clearance_concept(disposition) or {"text": disposition}
    obs = {"resourceType": "Observation", "_fullUrl": _urn(),
           "identifier": [{"system": config.SYS_REFERRAL_ID, "value": referral_id}],
           "status": "final",
           "code": {"coding": [{"system": config.SYS_LOINC,
                                "code": config.CLEARANCE_LOINC,
                                "display": "Referral clearance / disposition"}],
                    "text": "Referral clearance / disposition"},
           "valueCodeableConcept": concept}
    if patient_ref:
        obs["subject"] = {"reference": patient_ref}
    return obs


def build_medication_list(referral_id: str, patient_ref: str | None,
                          med_urls: list[str]) -> dict:
    """The ODEMedicationList (`List`) aggregating the patient's current medications.

    Per the ODE contract this is a base-FHIR `List` (status=current, mode=snapshot,
    code=LOINC 10160-0) whose entries reference the US Core MedicationRequests. It
    travels inside the referral Bundle and is referenced from ServiceRequest.supportingInfo.
    """
    lst = {"resourceType": "List", "_fullUrl": _urn(),
           "meta": {"profile": [config.PROFILE_ODE_MEDICATION_LIST]},
           "identifier": [{"system": config.SYS_REFERRAL_ID, "value": referral_id}],
           "status": "current", "mode": "snapshot",
           "title": "Medication List",
           "code": {"coding": [{"system": config.SYS_LOINC,
                                "code": config.MED_LIST_LOINC,
                                "display": "History of Medication use Narrative"}],
                    "text": "Medication List"},
           "entry": [{"item": {"reference": u}} for u in med_urls]}
    if patient_ref:
        lst["subject"] = {"reference": patient_ref}
    return lst


def task_output(resources: list[dict]) -> list[dict]:
    """Wrap result resources as Task.output entries (reference by fullUrl/id)."""
    out: list[dict] = []
    for res in resources or []:
        ref = res.get("_fullUrl") or (
            f"{res['resourceType']}/{res['id']}" if res.get("id") else None)
        if not ref:
            continue
        out.append({
            "type": {"text": res["resourceType"]},
            "valueReference": {"reference": ref},
        })
    return out


def apply_dental_profiles(resources: list[dict]) -> None:
    """Stamp provisional ODE dental profiles on dental-flavored resources in place."""
    for res in resources or []:
        rtype = res.get("resourceType")
        if rtype == "Procedure" and _is_cdt(res):
            _add_profile(res, config.PROFILE_DENTAL_PROCEDURE)
        elif rtype == "Observation" and (res.get("_dental_perio")
                                         or _profile_has(res, "perio")):
            _add_profile(res, config.PROFILE_PERIO_OBSERVATION)


def build_appointment(referral_id: str, patient_ref: str | None,
                      task_ref: str | None, start: str | None,
                      *, status: str = "booked", end: str | None = None,
                      location: str | None = None, provider: str | None = None,
                      appt_type: str | None = None) -> dict:
    """COW reply resource for PCC-60 Appointment Notification.

    Carries the scheduling detail a fulfiller sends: start/end, the booking
    provider (participant), a description (location) and appointmentType.
    """
    appt = {"resourceType": "Appointment", "_fullUrl": _urn(),
            "identifier": [{"system": config.SYS_REFERRAL_ID, "value": referral_id}],
            "status": status}
    if start:
        appt["start"] = _iso(start)
    if end:
        appt["end"] = _iso(end)
    type_cc = appointment_type_concept(appt_type)
    if type_cc:
        appt["appointmentType"] = type_cc
    if location:
        appt["description"] = location
    parts = []
    if patient_ref:
        parts.append({"actor": {"reference": patient_ref}, "status": "accepted"})
    if provider:
        parts.append({"actor": {"display": provider}, "status": "accepted"})
    if parts:
        appt["participant"] = parts
    if task_ref:
        appt["basedOn"] = [{"reference": task_ref}]
    return appt


def build_communication(referral_id: str, patient_ref: str | None,
                        task_ref: str | None, *, reason: str,
                        reason_code: str | None = None,
                        reschedule: str | None = None) -> dict:
    """COW reply resource for PCC-61 No-Show Notification.

    `reason_code` (optional) is a coded no-show reason; `reschedule` is a free-text
    reschedule instruction carried in the payload.
    """
    payloads = [{"contentString": reason}]
    if reschedule:
        payloads.append({"contentString": f"Reschedule: {reschedule}"})
    comm = {"resourceType": "Communication", "_fullUrl": _urn(),
            "identifier": [{"system": config.SYS_REFERRAL_ID, "value": referral_id}],
            "status": "completed",
            "reasonCode": [noshow_reason_concept(reason_code) or {"text": reason}],
            "payload": payloads}
    if patient_ref:
        comm["subject"] = {"reference": patient_ref}
    if task_ref:
        comm["partOf"] = [{"reference": task_ref}]
    return comm


def task_snapshot(referral_id: str, task_id: str | None, status: str,
                  business_status: str, *, focus_ref: str | None = None,
                  patient_ref: str | None = None,
                  outputs: list[dict] | None = None,
                  owner: str | dict | None = None,
                  status_reason: dict | None = None,
                  note: str | None = None,
                  period_end: str | None = None) -> dict:
    """A COW Task resource snapshot for the harness inbox / export (what the PMS
    would read from the FHIR server after the bridge applied a reply). Carries the
    reply content (owner/statusReason/note/restriction) when supplied."""
    task = {"resourceType": "Task",
            "meta": {"profile": [config.PROFILE_COW_TASK]},
            "identifier": [{"system": config.SYS_REFERRAL_ID, "value": referral_id}],
            "status": status, "intent": "order",
            "businessStatus": business_status_concept(business_status),
            "code": {"coding": [{
                "system": "http://hl7.org/fhir/CodeSystem/task-code",
                "code": "fulfill"}]}}
    if task_id:
        task["id"] = task_id
    if focus_ref:
        task["focus"] = {"reference": focus_ref}
    if patient_ref:
        task["for"] = {"reference": patient_ref}
    if owner:
        task["owner"] = {"reference": owner} if isinstance(owner, str) else owner
    if status_reason:
        task["statusReason"] = status_reason
    if note:
        task["note"] = [{"text": note}]
    if period_end:
        task["restriction"] = {"period": {"end": period_end}}
    if outputs:
        task["output"] = outputs
    return task


# --------------------------------------------------------------------------- #
def _iso(value: str) -> str:
    """Accept an HL7 v2 timestamp (YYYYMMDDHHMMSS) or pass through ISO-8601."""
    v = (value or "").strip()
    if v.isdigit() and len(v) >= 8:
        out = f"{v[0:4]}-{v[4:6]}-{v[6:8]}"
        if len(v) >= 12:
            out += f"T{v[8:10]}:{v[10:12]}:{v[12:14] or '00'}Z"
        return out
    return v


def _is_cdt(res: dict) -> bool:
    for c in (res.get("code", {}) or {}).get("coding", []):
        if c.get("system") == config.SYS_CDT:
            return True
    return False


def _profile_has(res: dict, needle: str) -> bool:
    return any(needle in p.lower()
               for p in (res.get("meta") or {}).get("profile", []))


def _add_profile(res: dict, profile: str) -> None:
    meta = res.setdefault("meta", {})
    profs = meta.setdefault("profile", [])
    if profile not in profs:
        profs.append(profile)
